File: server/src/users/dtos/user.dto.ts
import { Transform } from 'class-transformer';
import {
  IsArray,
  IsDefined,
  IsMongoId,
  IsNotEmpty,
  IsOptional,
  IsString,
} from 'class-validator';
import { AccessRole } from 'src/common/types';

export class UserDto {
  @IsMongoId()
  id: string;

  @IsMongoId()
  rootId: string;

  @IsString()
  @IsNotEmpty()
  @Transform(({ value }) => value.toLowerCase(), { toClassOnly: true })
  name: string;

  @IsString()
  @IsNotEmpty()
  @Transform(({ value }) => value.toLowerCase(), { toClassOnly: true })
  last: string;

  @IsString()
  @IsNotEmpty()
  @Transform(({ value }) => value.toLowerCase(), { toClassOnly: true })
  username: string;

  @IsString()
  @IsNotEmpty()
  password: string;

  @IsArray()
  @IsDefined()
  access: string[];

  @IsArray()
  @IsOptional()
  roles?: AccessRole[];
}

File: clients/flutter_client/lib/src/http/dtos/common/user_dto.dart
class UserDto {
  final String id;
  final String rootId;
  final String name;
  final String last;
  final String username;
  final List<String> access;
  final List<String> roles;

  const UserDto(
      {required this.id,
      required this.rootId,
      required this.name,
      required this.last,
      required this.username,
      required this.access,
      required this.roles});

  factory UserDto.fromJson(Map<String, dynamic> json) {
    return UserDto(
        id: json['id'],
        rootId: json['rootId'],
        name: json['name'],
        last: json['last'],
        username: json['username'],
        access: List<String>.from(json['access']),
        roles: List<String>.from(json['roles']));
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'rootId': rootId,
      'name': name,
      'last': last,
      'username': username,
      'access': access,
      'roles': roles,
    };
  }
}
